src/blr/cli/tagbam.py

```diff
@@ -1,58 +1,100 @@
+Transfers SAM tags from query headers to SAM tags. Currently tags in header must follow SAM tag format, e.g.
+BC:Z:<SEQUENCE>.
 """
 
 import pysam
 import logging
+import re
 from tqdm import tqdm
+from collections import Counter
+
+from blr.utils import print_stats
 
 logger = logging.getLogger(__name__)
 
+ALLOWED_SAM_TAG_TYPES = "ABfHiZ"  # From SAM format specs https://samtools.github.io/hts-specs/SAMtags.pdf
+
 
 def main(args):
+    # Can't be at top since function are defined later
+    REGEX_FUNCTION_DICT = {
+        "sam": build_regex_sam_tag
+    }
 
-    # Generate dict with bc => bc_cluster consensus sequence
     logger.info("Starting analysis")
-    alignments_missing_bc = 0
+    summary = Counter()
+    regex_function = REGEX_FUNCTION_DICT[args.pattern_type]
 
     # Read SAM/BAM files and transfer barcode information from alignment name to SAM tag
     with pysam.AlignmentFile(args.input, "rb") as infile, \
             pysam.AlignmentFile(args.output, "wb", template=infile) as out:
-
         for read in tqdm(infile.fetch(until_eof=True), desc="Reading input"):
-            try:
-                name, raw_barcode_tag, corr_barcode_tag = read.query_name.split()[0].split('_')
-            except ValueError:
-                alignments_missing_bc += 1
-                name, raw_barcode_tag, corr_barcode_tag = read.query_name.split()[0], None, None
+            for tag in args.tags:
 
-            # Rename aligment to exclude barcode information.
-            read.query_name = name
+                # Make regex expression and search for tag
+                pattern = regex_function(bam_tag=tag)
+                match = re.search(pattern, read.query_name)
 
-            if corr_barcode_tag:
-                assert corr_barcode_tag.startswith(f"{args.barcode_tag}:Z:")
-                corr_barcode = corr_barcode_tag.split(":")[-1]
-                read.set_tag(args.barcode_tag, corr_barcode, value_type='Z')
-
-            if raw_barcode_tag:
-                assert raw_barcode_tag.startswith(f"{args.sequence_tag}:Z:")
-                raw_barcode = raw_barcode_tag.split(":")[-1]
-                read.set_tag(args.sequence_tag, raw_barcode, value_type='Z')
+                # If tag string is found, remove from header and set SAM tag value
+                if match:
+                    full_tag_string = match.group(0)
+                    divider = read.query_name[match.start() - 1]
+                    read.query_name = read.query_name.replace(divider + full_tag_string, "")
+                    if not args.only_remove:
+                        read.set_tag(tag, match.group("value"), value_type=match.group("type"))
+                    summary[f"reads with tag {tag}"] += 1
+                else:
+                    summary[f"reads without tag {tag}"] += 1
 
             out.write(read)
 
-    logger.info(f"Alignments missing barcodes: {alignments_missing_bc}")
+    print_stats(summary, name="stats")
```

Comment:
```suggestion
    print_stats(summary, name=__name__)
```
In the other scripts I have included the script name in the summary table